Go workspaces for k/k and k/staging/*

thockin opened the pull request:
KEP: https://github.com/kubernetes/enhancements/issues/4402

This is a new PR based on #122624, to clear the lengthy history.

What is it?

The main Kubernetes git repo (github.com/kubernetes/kubernetes, colloquially called "k/k") is the embodiment of evolution.  When it was created, the only way to build Go applications was GOPATH.  Over time we built tooling which understood GOPATH, and it leaked into many parts of our build, test, and release systems.  Then Go added modules, in part because GOPATH was unpleasant and had a tendency to leak into tools.  We adopted modules, but we also added the idea of "staging" repositories - a way to eat our cake and have, too.  We get the benefits of a monorepo (atomic commits, fast iteration across repos) and then publish those to standalone repos for downstream consumption.  To make this work with modules, we abused GOPATH even harder and wrote even more tools.

The Go project saw what we (and others) were doing and did not like it.  So they created [workspaces](https://go.dev/doc/tutorial/workspaces).  They basically created a solution that is purpose-built for us.

Let's use it.

On reviewing:  Review each commit individually.  This PR was crafted in lock-step with a series of gengo commits, but since those are merged and gengo is in a different repo, we can't really see those any more.  This means that a lot of things don't work at each commit in this PR.  This is the only way this could be a sane PR to review though (if you call this sane).

I fully expect that once this merges we will find a long-tail of out-of-tree impacts (tests which need to use go 1.22, etc).

One major aspect of this is code-generation.  It's FAR simpler now, but how does it perform?  After factoring out build time (by doing a full build first) we can see it has little impact:

before:
```
$ time ./hack/update-codegen.sh protobuf deepcopy prerelease defaults conversions openapi applyconfigs clients listers informers
go version go1.21.5 linux/amd64
+++ [0107 16:21:01] Generating protobufs for 67 targets
+++ [0107 16:21:01] protoc 23.4 not found (can install with hack/install-protoc.sh); generating containerized...
+++ [0107 16:21:01] Verifying Prerequisites....
+++ [0107 16:21:02] Building Docker image kube-build:build-d19b652805-5-v1.29.0-go1.21.5-bullseye.0
+++ [0107 16:21:04] Syncing sources to container
+++ [0107 16:21:06] Output from this container will be rsynced out upon completion. Set KUBE_RUN_COPY_OUTPUT=n to disable.
+++ [0107 16:21:06] Running build command...
+++ [0107 16:21:25] Syncing out of container
+++ [0107 16:21:27] Generating deepcopy code for 250 targets
+++ [0107 16:21:32] Generating prerelease-lifecycle code for 30 targets
+++ [0107 16:21:34] Generating defaulter code for 100 targets
+++ [0107 16:21:42] Generating conversion code for 139 targets
+++ [0107 16:22:19] Generating openapi code
+++ [0107 16:22:29] Generating apply-config code for 56 targets
+++ [0107 16:22:35] Generating client code for 51 targets
+++ [0107 16:22:40] Generating lister code for 55 targets
+++ [0107 16:22:42] Generating informer code for 55 targets

real	1m43.608s
user	1m29.445s
sys	0m23.746s

```

after:
```
$ time ./hack/update-codegen.sh protobuf deepcopy prerelease defaults conversions openapi applyconfigs clients listers informers
+++ [0107 16:15:22] Generating protobufs for 67 targets
+++ [0107 16:15:23] protoc 23.4 not found (can install with hack/install-protoc.sh); generating containerized...
+++ [0107 16:15:23] Verifying Prerequisites....
+++ [0107 16:15:23] Building Docker image kube-build:build-08a5207a41-5-v1.29.0-go1.21.5-bullseye.0
+++ [0107 16:15:25] Syncing sources to container
+++ [0107 16:15:27] Output from this container will be rsynced out upon completion. Set KUBE_RUN_COPY_OUTPUT=n to disable.
+++ [0107 16:15:27] Running build command...
+++ [0107 16:15:49] Syncing out of container
+++ [0107 16:15:51] Generating deepcopy code for 250 targets
+++ [0107 16:15:57] Generating prerelease-lifecycle code for 30 targets
+++ [0107 16:15:59] Generating defaulter code for 100 targets
+++ [0107 16:16:04] Generating conversion code for 139 targets
+++ [0107 16:16:15] Generating openapi code
+++ [0107 16:16:26] Generating apply-config code for 56 targets
+++ [0107 16:16:33] Generating client code for 51 targets
+++ [0107 16:16:39] Generating lister code for 55 targets
+++ [0107 16:16:42] Generating informer code for 55 targets

real	1m23.480s
user	1m49.018s
sys	0m21.897s
```

/kind bug
/kind cleanup
/kind feature
/kind api-change


```release-note
The kubernetes repo now uses Go workspaces.  This should not impact end users at all, but does have impact for developers of downstream projects.  Switching to workspaces caused some breaking changes in the flags to the various k8s.io/code-generator tools.  Downstream consumers should look at staging/src/k8s.io/code-generator/kube_codegen.sh to see the changes.
```

thockin committed:
Workspaces: excise vendor from golang.sh

```
$ make
go version go1.20.4 linux/amd64
+++ [0506 13:21:33] Building go targets for linux/amd64
    ./cmd/kube-proxy (static)
    ./cmd/kube-apiserver (static)
    ./cmd/kube-controller-manager (static)
    ./cmd/kubelet (non-static)
    ./cmd/kubeadm (static)
    ./cmd/kube-scheduler (static)
    ./staging/src/k8s.io/component-base/logs/kube-log-runner (static)
    ./staging/src/k8s.io/kube-aggregator (static)
    ./staging/src/k8s.io/apiextensions-apiserver (static)
    ./cluster/gce/gci/mounter (non-static)
    ./cmd/kubectl (static)
    ./cmd/kubectl-convert (static)
    github.com/onsi/ginkgo/v2/ginkgo (non-static)
    ./test/e2e/e2e.test (test)
    ./test/conformance/image/go-runner (non-static)
    ./cmd/kubemark (static)
    github.com/onsi/ginkgo/v2/ginkgo (non-static)
    ./test/e2e_node/e2e_node.test (test)
```

before:
real	4m31.100s
user	21m1.466s
sys	2m23.546s

after:
real	3m46.591s
user	16m34.665s
sys	2m1.553s

thockin committed:
Remove vendor from ldflags

thockin committed:
Clean up test.sh and get rid of GOPATH

Result: `make test` works:

```
$ #### BUILD ####

$ make WHAT="./cmd/kubectl/ ./staging/src/k8s.io/api"
+++ [1211 11:20:46] Building go targets for linux/amd64
    ./cmd/kubectl/ (non-static)
    ./staging/src/k8s.io/api (non-static)

$ #### TEST ####

$ make test WHAT=./pkg/proxy/iptables/
+++ [1211 11:22:31] Set GOMAXPROCS automatically to 6
+++ [1211 11:22:31] Running tests without code coverage and with -race
ok  	k8s.io/kubernetes/pkg/proxy/iptables	9.517s

$ make test WHAT=pkg/proxy/iptables/
+++ [1211 11:22:53] Set GOMAXPROCS automatically to 6
+++ [1211 11:22:53] Running tests without code coverage and with -race
ok  	k8s.io/kubernetes/pkg/proxy/iptables	(cached)

$ make test WHAT=k8s.io/kubernetes/pkg/proxy/iptables/
+++ [1211 11:23:09] Set GOMAXPROCS automatically to 6
+++ [1211 11:23:09] Running tests without code coverage and with -race
ok  	k8s.io/kubernetes/pkg/proxy/iptables	(cached)

$ make test WHAT=./staging/src/k8s.io/api
+++ [1211 11:23:24] Set GOMAXPROCS automatically to 6
+++ [1211 11:23:24] Running tests without code coverage and with -race
ok  	k8s.io/api	21.981s

$ make test WHAT=staging/src/k8s.io/api
+++ [1211 11:23:54] Set GOMAXPROCS automatically to 6
+++ [1211 11:23:54] Running tests without code coverage and with -race
ok  	k8s.io/api	(cached)

$ make test WHAT=k8s.io/api
+++ [1211 11:24:06] Set GOMAXPROCS automatically to 6
+++ [1211 11:24:06] Running tests without code coverage and with -race
ok  	k8s.io/api	(cached)
```

thockin committed:
Remove some references to GOPATH

thockin committed:
Fix update/verify-mocks.sh

There appears to be a bug in `go generate` for workspaces which will be
fixed in the 1.22.1 release.

thockin committed:
Change the build_binaries path to use modules

This makes "new" and "old" setup_env functions. In subsequent commits,
all callers of the "old" form will be fixed, and the "new" will be
renamed back.

The old and new functions diff:

```diff
--- /tmp/a	2023-12-14 09:02:57.804092696 -0800
+++ /tmp/b	2023-12-14 09:03:09.679999585 -0800
@@ -1,4 +1,4 @@
-kube::golang::old::setup_env() {
+kube::golang::new::setup_env() {
   kube::golang::verify_go_version

   # Set up GOPATH.  We have tools which depend on being in a GOPATH (see
@@ -7,9 +7,9 @@
   # Even in module mode, we need to set GOPATH for `go build` and `go install`
   # to work.  We build various tools (usually via `go install`) from a lot of
   # scripts.
-  #   * We can't set GOBIN because that does not work on cross-compiles.
-  #   * We could use `go build -o <something>`, but it's subtle when it comes
-  #     to cross-compiles and whether the <something> is a file or a directory,
+  #   * We can't just set GOBIN because that does not work on cross-compiles.
+  #   * We could always use `go build -o <something>`, but it's subtle wrt
+  #     cross-compiles and whether the <something> is a file or a directory,
   #     and EVERY caller has to get it *just* right.
   #   * We could leave GOPATH alone and let `go install` write binaries
   #     wherever the user's GOPATH says (or doesn't say).
@@ -20,16 +20,6 @@
   #
   # Eventually, when we no longer rely on run-in-gopath.sh we may be able to
   # simplify this some.
-  local go_pkg_dir="${KUBE_GOPATH}/src/${KUBE_GO_PACKAGE}"
-  local go_pkg_basedir
-  go_pkg_basedir=$(dirname "${go_pkg_dir}")
-
-  mkdir -p "${go_pkg_basedir}"
-
-  # TODO: This symlink should be relative.
-  if [[ ! -e "${go_pkg_dir}" || "$(readlink "${go_pkg_dir}")" != "${KUBE_ROOT}" ]]; then
-    ln -snf "${KUBE_ROOT}" "${go_pkg_dir}"
-  fi
   export GOPATH="${KUBE_GOPATH}"

   # If these are not set, set them now.  This ensures that any subsequent
@@ -40,24 +30,10 @@
   # Make sure our own Go binaries are in PATH.
   export PATH="${KUBE_GOPATH}/bin:${PATH}"

-  # Change directories so that we are within the GOPATH.  Some tools get really
-  # upset if this is not true.  We use a whole fake GOPATH here to collect the
-  # resultant binaries.
-  local subdir
-  subdir=$(kube::realpath . | sed "s|${KUBE_ROOT}||")
-  cd "${KUBE_GOPATH}/src/${KUBE_GO_PACKAGE}/${subdir}" || return 1
-
-  # Set GOROOT so binaries that parse code can work properly.
-  GOROOT=$(go env GOROOT)
-  export GOROOT
-
   # Unset GOBIN in case it already exists in the current session.
   # Cross-compiles will not work with it set.
   unset GOBIN

-  # This seems to matter to some tools
-  export GO15VENDOREXPERIMENT=1
-
-  # Disable workspaces
-  export GOWORK=off
+  # Explicitly turn on modules.
+  export GO111MODULE=on
 }
```

Result: `make` works for k/k:

```
$ make kubectl
+++ [1211 11:07:31] Building go targets for linux/amd64
    k8s.io/kubernetes/cmd/kubectl (static)

$ make WHAT=./cmd/kubectl/
+++ [1211 11:08:19] Building go targets for linux/amd64
    k8s.io/kubernetes/./cmd/kubectl/ (non-static)

$ make WHAT=k8s.io/kubernetes/cmd/kubectl
+++ [1211 11:08:52] Building go targets for linux/amd64
    k8s.io/kubernetes/cmd/kubectl (static)
```

Result: `make` works for staging by package:

```
$ make WHAT=k8s.io/api
+++ [1211 11:11:37] Building go targets for linux/amd64
    k8s.io/api (non-static)
```

Result: `make` fails for staging by path:

```
$ make WHAT=./staging/src/k8s.io/api
+++ [1211 11:12:44] Building go targets for linux/amd64
    k8s.io/kubernetes/./staging/src/k8s.io/api (non-static)
cannot find module providing package k8s.io/kubernetes/staging/src/k8s.io/api: import lookup disabled by -mod=vendor
	(Go version in go.work is at least 1.14 and vendor directory exists.)
!!! [1211 11:12:44] Call tree:
!!! [1211 11:12:44]  1: /home/thockin/src/kubernetes/hack/lib/golang.sh:850 kube::golang::build_some_binaries(...)
!!! [1211 11:12:44]  2: /home/thockin/src/kubernetes/hack/lib/golang.sh:1012 kube::golang::build_binaries_for_platform(...)
!!! [1211 11:12:44]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [1211 11:12:44] Call tree:
!!! [1211 11:12:44]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [1211 11:12:44] Call tree:
!!! [1211 11:12:44]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make: *** [Makefile:96: all] Error 1
```

Result: `make test` fails:

```
$ make test WHAT=./cmd/kubectl
+++ [1211 11:13:38] Set GOMAXPROCS automatically to 6
+++ [1211 11:13:38] Running tests without code coverage and with -race
cmd/kubectl/kubectl.go:25:2: cannot find package "k8s.io/client-go/plugin/pkg/client/auth" in any of:
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/plugin/pkg/client/auth (vendor tree)
	/home/thockin/sdk/gotip/src/k8s.io/client-go/plugin/pkg/client/auth (from $GOROOT)
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/client-go/plugin/pkg/client/auth (from $GOPATH)
cmd/kubectl/kubectl.go:20:2: cannot find package "k8s.io/component-base/cli" in any of:
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/component-base/cli (vendor tree)
	/home/thockin/sdk/gotip/src/k8s.io/component-base/cli (from $GOROOT)
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/component-base/cli (from $GOPATH)
cmd/kubectl/kubectl.go:21:2: cannot find package "k8s.io/kubectl/pkg/cmd" in any of:
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd (vendor tree)
	/home/thockin/sdk/gotip/src/k8s.io/kubectl/pkg/cmd (from $GOROOT)
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubectl/pkg/cmd (from $GOPATH)
cmd/kubectl/kubectl.go:22:2: cannot find package "k8s.io/kubectl/pkg/cmd/util" in any of:
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/util (vendor tree)
	/home/thockin/sdk/gotip/src/k8s.io/kubectl/pkg/cmd/util (from $GOROOT)
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubectl/pkg/cmd/util (from $GOPATH)
make: *** [Makefile:191: test] Error 1
```

thockin committed:
Excise KUBE_GOPATH from coverage build

thockin committed:
Rename KUBE_OUTPUT_BINPATH -> KUBE_OUTPUT_BIN

"...PATH" doesn't add anything and confuses it with KUBE_OUTPUT_SUBPATH.

thockin committed:
Clarify, document KUBE_OUTPUT_* variables

thockin committed:
Change verify-api-groups.sh to s/vendor/staging/

I'm really not sure what this particular file check is doing, but that's
for tomorrow.

thockin committed:
Simplify verify-cli-conventions; use new setup_env

pohly committed:
Generate go.work files

This creates go.work and enables Go Workspaces.  This is a file that
includes info on k/k and all the staging modules.

This depends on go 1.22 and setting FORCE_HOST_GO=true (for kube
scripts, which try to be hermetic).

Make this part of the normal update/verify sequence.

The top-level go.work file contains no replace statements. Instead, the
replace statements in the individual go.mod files are used. For this to
work, replace statements in the individual go.mod files have to be
consistent.

hack/tools has different dependencies and can't be in the main
workspace, so this adds a go.work just for that.  Without this, go tries
to consider all deps in all modules and pick one that works for all.
This is problematic because there are so many of them that it is
difficult to manage.

Likewise for k8s.io/code-generator/examples and
k8s.io/kms/internal/plugins/_mock - add trivial go.work files.

For example k/k depends on an older version of a lib that gloangci-lint
needs (transitively) and it breaks.

This also updates vendor (needed to make go happy), and removes
vendor'ed symlinks.  This breaks a LOT of our build tools, which will be
fixed subsequently.

Result: `go` commands work across modules:

Before:
```
$ go list ./pkg/proxy/iptables/ ./staging/src/k8s.io/api/core/v1/
main module (k8s.io/kubernetes) does not contain package k8s.io/kubernetes/staging/src/k8s.io/api/core/v1

$ go build ./pkg/proxy/iptables/ ./staging/src/k8s.io/api
main module (k8s.io/kubernetes) does not contain package k8s.io/kubernetes/staging/src/k8s.io/api

$ go test ./pkg/proxy/iptables/ ./staging/src/k8s.io/api
main module (k8s.io/kubernetes) does not contain package k8s.io/kubernetes/staging/src/k8s.io/api
```

After:
```
$ go list ./pkg/proxy/iptables/ ./staging/src/k8s.io/api/core/v1/
k8s.io/kubernetes/pkg/proxy/iptables
k8s.io/api/core/v1

$ go build ./pkg/proxy/iptables/ ./staging/src/k8s.io/api

$ go test ./pkg/proxy/iptables/ ./staging/src/k8s.io/api
ok  	k8s.io/kubernetes/pkg/proxy/iptables	0.360s
ok  	k8s.io/api	2.302s
```

Result: `make` fails:

```
$ make
go version go1.22rc1 linux/amd64
+++ [0106 12:11:03] Building go targets for linux/amd64
    k8s.io/kubernetes/cmd/kube-proxy (static)
    k8s.io/kubernetes/cmd/kube-apiserver (static)
    k8s.io/kubernetes/cmd/kube-controller-manager (static)
    k8s.io/kubernetes/cmd/kubelet (non-static)
    k8s.io/kubernetes/cmd/kubeadm (static)
    k8s.io/kubernetes/cmd/kube-scheduler (static)
    k8s.io/component-base/logs/kube-log-runner (static)
    k8s.io/kube-aggregator (static)
    k8s.io/apiextensions-apiserver (static)
    k8s.io/kubernetes/cluster/gce/gci/mounter (static)
    k8s.io/kubernetes/cmd/kubectl (static)
    k8s.io/kubernetes/cmd/kubectl-convert (static)
    github.com/onsi/ginkgo/v2/ginkgo (non-static)
    k8s.io/kubernetes/test/e2e/e2e.test (test)
    k8s.io/kubernetes/test/conformance/image/go-runner (non-static)
    k8s.io/kubernetes/cmd/kubemark (static)
    github.com/onsi/ginkgo/v2/ginkgo (non-static)
    k8s.io/kubernetes/test/e2e_node/e2e_node.test (test)
test/e2e/e2e.go:35:2: cannot find package "k8s.io/api/apps/v1" in any of:
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/api/apps/v1 (vendor tree)
	/home/thockin/src/kubernetes/_output/local/.gimme/versions/go1.22rc1.linux.amd64/src/k8s.io/api/apps/v1 (from $GOROOT)
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/api/apps/v1 (from $GOPATH)
	... more ...
	... more ...
	... more ...
!!! [0106 12:13:41] Call tree:
!!! [0106 12:13:41]  1: /home/thockin/src/kubernetes/hack/lib/golang.sh:948 kube::golang::build_binaries_for_platform(...)
!!! [0106 12:13:41]  2: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0106 12:13:41] Call tree:
!!! [0106 12:13:41]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0106 12:13:41] Call tree:
!!! [0106 12:13:41]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make: *** [Makefile:96: all] Error 1
```

Again, this requires go 1.22 (e.g. gotip), as go 1.21.x does not have
`go work vendor` support.

TO REPEAT:
    ( \
      ./hack/update-go-workspace.sh; \
      ./hack/update-vendor.sh; \
      ./hack/update-go-workspace.sh; \
    )

thockin committed:
Make golang::binaries_from_targets smarter

and use it from test

Result: `make` works, but `make test` still broken

```
$ make kubectl
+++ [1211 11:15:46] Building go targets for linux/amd64
    ./cmd/kubectl (static)

$ make WHAT=./cmd/kubectl/
+++ [1211 11:15:53] Building go targets for linux/amd64
    ./cmd/kubectl/ (non-static)

$ make WHAT=cmd/kubectl/
+++ [1211 11:16:02] Building go targets for linux/amd64
    ./cmd/kubectl/ (non-static)

$ make WHAT=k8s.io/kubernetes/cmd/kubectl
+++ [1211 11:16:16] Building go targets for linux/amd64
    k8s.io/kubernetes/cmd/kubectl (static)

$ make WHAT=./staging/src/k8s.io/api
+++ [1211 11:16:31] Building go targets for linux/amd64
    ./staging/src/k8s.io/api (non-static)

$ make WHAT=staging/src/k8s.io/api
+++ [1211 11:16:43] Building go targets for linux/amd64
    ./staging/src/k8s.io/api (non-static)

$ make WHAT=k8s.io/api
+++ [1211 11:16:50] Building go targets for linux/amd64
    k8s.io/api (non-static)
```

```
$ make test WHAT=./cmd/kubectl
+++ [1211 11:17:23] Set GOMAXPROCS automatically to 6
+++ [1211 11:17:23] Running tests without code coverage and with -race
cmd/kubectl/kubectl.go:25:2: cannot find package "k8s.io/client-go/plugin/pkg/client/auth" in any of:
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/plugin/pkg/client/auth (vendor tree)
	/home/thockin/sdk/gotip/src/k8s.io/client-go/plugin/pkg/client/auth (from $GOROOT)
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/client-go/plugin/pkg/client/auth (from $GOPATH)
cmd/kubectl/kubectl.go:20:2: cannot find package "k8s.io/component-base/cli" in any of:
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/component-base/cli (vendor tree)
	/home/thockin/sdk/gotip/src/k8s.io/component-base/cli (from $GOROOT)
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/component-base/cli (from $GOPATH)
cmd/kubectl/kubectl.go:21:2: cannot find package "k8s.io/kubectl/pkg/cmd" in any of:
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd (vendor tree)
	/home/thockin/sdk/gotip/src/k8s.io/kubectl/pkg/cmd (from $GOROOT)
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubectl/pkg/cmd (from $GOPATH)
cmd/kubectl/kubectl.go:22:2: cannot find package "k8s.io/kubectl/pkg/cmd/util" in any of:
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/util (vendor tree)
	/home/thockin/sdk/gotip/src/k8s.io/kubectl/pkg/cmd/util (from $GOROOT)
	/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubectl/pkg/cmd/util (from $GOPATH)
make: *** [Makefile:191: test] Error 1
```

thockin committed:
Dedup binaries when building

thockin committed:
Make build scripts use Go import paths

This helps 2 things:
1) output is cleaner
2) whether a build is static or non-static depends on string matching,
   which is easier after this

If you specify a `...` pattern, this will preserve it, which breaks the
static/non-static matching but really, can we just agree that `make` is
for binaries?  It works for other things but so does `go build`.

Example:

```
$ make WHAT="cmd/kubectl/ ./staging/src/k8s.io/api/... ./test/e2e/e2e.test ./test/e2e/e2e.test ./cmd/kubectl/ "
go version go1.21.4 linux/amd64
+++ [1208 22:24:43] Building go targets for linux/amd64
    k8s.io/api/... (non-static)
    k8s.io/kubernetes/cmd/kubectl (static)
    k8s.io/kubernetes/test/e2e/e2e.test (test)
```

thockin committed:
Change update-openapi-spec to new setup_env

thockin committed:
Change verify-licenses to new setup_env

thockin committed:
Change verify-gofmt to new setup_env

thockin committed:
Change verify-yamlfmt to new setup_env

thockin committed:
Change update-golangci-lint-config to new setup_env

thockin committed:
Change verify-description to new setup_env

thockin committed:
Change verify-fieldname-docs to new setup_env

thockin committed:
Change verify-generated-docs to new setup_env

thockin committed:
Change verify-govulncheck to new setup_env

thockin committed:
Change test/instrumentation/update-* to new setup_env

thockin committed:
Change ipamperf test to new setup_env

thockin committed:
Change update-generated-api-compatibility-data test to new setup_env

thockin committed:
Make verify-no-vendor-cycles use new setup_env

thockin committed:
Make verify-conformance-requirements use new setup_env

thockin committed:
Change verify-generated to new setup_env

thockin committed:
Change verify-imports to use new setup_env

Also clean up the tool a bit and drop vendor as an explicit thing to
be handled.

thockin committed:
Change test-e2e-node to use new setup_env

thockin committed:
Fix lister-gen wrt gengo/v2

thockin committed:
Fix informer-gen wrt gengo/v2

thockin committed:
Fix applyconfiguration-gen wrt gengo/v2

thockin committed:
Fix client-gen wrt gengo/v2

thockin committed:
Fix conversion-gen wrt gengo/v2

thockin committed:
Fix prerelease-lifecycle-gen wrt gengo/v2

thockin committed:
Fix register-gen wrt gengo/v2

thockin committed:
Fix go-to-protobuf wrt gengo/v2

There's some very fishy-smelling logic in here, but this commit is
trying to be as focused as possible.

The *.pb.go diffs are the "name" encoded in the descriptor.  The
descriptor blobs can be decoded by this program (thanks StackOverflow!):

```
package main

import (
	"bytes"
	"compress/gzip"
	"encoding/json"
	"fmt"
	"os"

	"io/ioutil"

	proto "github.com/golang/protobuf/proto"
	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
)

func main() {
	m := map[string][]byte{
		"before": blobv1,
		"after":  blobv2,
	}
	arg := os.Args[1]
	dump(m[arg])
}

func dump(bytes []byte) {
	fd, err := decodeFileDesc(bytes)
	if err != nil {
		panic(err)
	}
	b, err := json.MarshalIndent(fd, "", "  ")
	if err != nil {
		panic(err)
	}
	fmt.Println(string(b))
}

// decompress does gzip decompression.
func decompress(b []byte) ([]byte, error) {
	r, err := gzip.NewReader(bytes.NewReader(b))
	if err != nil {
		return nil, fmt.Errorf("bad gzipped descriptor: %v", err)
	}
	out, err := ioutil.ReadAll(r)
	if err != nil {
		return nil, fmt.Errorf("bad gzipped descriptor: %v", err)
	}
	return out, nil
}

func decodeFileDesc(enc []byte) (*dpb.FileDescriptorProto, error) {
	raw, err := decompress(enc)
	if err != nil {
		return nil, fmt.Errorf("failed to decompress enc: %v", err)
	}

	fd := new(dpb.FileDescriptorProto)
	if err := proto.Unmarshal(raw, fd); err != nil {
		return nil, fmt.Errorf("bad descriptor: %v", err)
	}
	return fd, nil
}

var blobv1 = []byte{
	// insert proto "before" blob here
}

var blobv2 = []byte{
	// insert proto "after" blob here
}
```

Running this with "before" and "after" args, and diffing the output
yields something like:

```diff
--- /tmp/a	2023-12-23 23:57:04.748090836 -0800
+++ /tmp/b	2023-12-23 23:57:11.000040973 -0800
@@ -1,5 +1,5 @@
 {
-  "name": "k8s.io/kubernetes/vendor/k8s.io/api/admission/v1/generated.proto",
+  "name": "k8s.io/api/admission/v1/generated.proto",
   "package": "k8s.io.api.admission.v1",
   "dependency": [
     "github.com/gogo/protobuf/gogoproto/gogo.proto",
```

thockin committed:
Change update-codegen.sh to use new setup_env

This breaks all the generators, which will be fixed subsequently.

thockin committed:
Make update-codegen protobuf work on gengo/v2

thockin committed:
Make update-codegen deepcopy work on gengo/v2

thockin committed:
Make update-codegen swagger work on gengo/v2

thockin committed:
Make update-codegen prerelease work on gengo/v2

thockin committed:
Make update-codegen applyconfig work on gengo/v2

thockin committed:
Make update-codegen openapi work on gengo/v2

thockin committed:
Make update-codegen client work on gengo/v2

thockin committed:
Make update-codegen lister work on gengo/v2

thockin committed:
Make update-codegen informer work on gengo/v2

thockin committed:
Make code-gen subprojects work on gengo/v2

thockin committed:
Fix import-boss wrt gengo/v2

thockin committed:
Fix last run-in-gopath user: typecheck

This now works across modules

Verified by introducing errors in 2 different modules and running.

thockin committed:
Remove run-in-gopath.sh

thockin committed:
Retool typecheck to be simpler

Instead of walking paths ourselves, just let Go's packages library do
it.  This is a slight CLI change - it wants "./foo" rather than "foo".

This also flagged a few things which seem to be legit failures.

thockin committed:
Remove old::setup_env

thockin committed:
Rename new::setup_env back to setup_env

thockin committed:
Make update-codegen defaulter work on gengo/v2

thockin committed:
Make update-codegen conversion work on gengo/v2

thockin committed:
Kill off PRJ_SRC_PATH

thockin committed:
code-generator: Remove now unused helper function

thockin committed:
Get rid of most references to GOPATH

thockin committed:
Update Makefile comments

thockin committed:
Fix update-import-aliases for workspaces

thockin committed:
Fix lint-dependencies for workspaces

thockin committed:
Fix update-netparse-cve for workspaces

thockin committed:
Fix verify-external-dependencies-version for workspaces

thockin committed:
Fix verify-spelling for workspaces

thockin committed:
Fix pin-dependency for workspaces

thockin committed:
Tidy update-vendor-licenses for workspaces

thockin committed:
Remove references to "GO111MODULE"

thockin committed:
Call setup_env from verify-vendor

thockin committed:
Make update-vendor log the go env

thockin committed:
Require Go 1.22.0 in go.mod

thockin committed:
Conversion: Better logging

thockin committed:
codegen: Use single-dash for '-v'

thockin committed:
Call verify_go_version at the END of setup_env

Prior to this it would download Go, setup the environment, then (maybe)
download Go again.

I renamed verify_go_version to be internal and fixed all callers.

thockin committed:
Rename update-generated-* dockerized

Now they don't start with "update" and confuse dumb scripts which try to
run all the updates.

thockin committed:
Fix verify-conformance-yaml for workspaces

thockin committed:
Fix scripts to avoid testdata

thockin committed:
Fix golangci-lint for workspaces

thockin committed:
Make verify-shellcheck take optional args

thockin committed:
Fix make test-integration for workspaces

thockin committed:
Remove KUBE_GO_PACKAGE

Almost nobody should need it, and if you do, it should be an immediate
smell.

thockin committed:
codegen: Don't need to pass --logtostderr

All the tools already set it internally

thockin committed:
Get rid of vendor baggage in go-to-proto

thockin committed:
codegen: Minor: fix wrong comments

thockin committed:
codegen: Make import-boss do it's own flags

thockin committed:
Move import-boss to k/k, retool to not use gengo

* Moved code and tests out of gengo -> code_generator
* Reworked it to use packages.Load
* Reworked tests (still not comprehensive but pretty good?)
* Dropped test support from gengo (support for tests in
  x/tools/go/packages is pretty hostile to gengo, and nobody used it)

thockin committed:
codegen: Remove most uses of InputDirs

This clears the path to getting rid of it completely.

thockin committed:
Move gengo/examples/deepcopy -> code_generator

Merge comments in main.go and fix a bunch of pre-existing lint errors.

thockin committed:
Move gengo/examples/defaulter -> code_generator

This includes tests, which all pass.  Also fixes a bunch of lint.

thockin committed:
Remove legacy references to './output'

thockin committed:
Move doc-generation funcs into only caller

thockin committed:
Make hack scripts use `go install` and assume PATH

Now that they all call setup_env, we don't need find-binary (I think).
That was originally meant to hide the diff between docker and local
builds but all these tools do local builds anyway.

thockin committed:
Fix lint found by golangci-lint

thockin committed:
Remove dep on set-gen: gone in gengo/v2

thockin committed:
Convert all gengo references to gengo/v2

TO REPEAT:

    git ls-files ':!:vendor/k8s.io/gengo' ':!:gengo-v2' ':(glob)**/*.go' \
        | xargs grep -l 'k8s.io/gengo/' \
        | while read F; do \
            sed -i 's|k8s.io/gengo/\([^v]\)|k8s.io/gengo/v2/\1|' $F
        done

thockin committed:
Remove defunct references to "vendor"

thockin committed:
Remove defunct trim-path-prefix

thockin committed:
codegen:proto: Use revised API to gengo parser

thockin committed:
Rename DefaultPackage to SimplePackage

thockin committed:
Drop `generator.Packages` type

thockin committed:
Rename gengo generator.Package -> Target

thockin committed:
Rename gengo Target.SourcePath() -> Dir()

thockin committed:
Rename gengo DefaultGen -> GolangGenerator

thockin committed:
Type-assert gengo CustomArgs blindly

If it does fail, it is catastophic.  Don't bother trying to handle it.

thockin committed:
Move the OutputBase flag to each tool

...and rename to --output-dir

thockin committed:
Rename output-package -> output-pkg

To parallel "output-dir".

thockin committed:
Move OutputFileBase flag to each tool

thockin committed:
Remove gengo verify-only - unused and untested

All of our own tools emit into a tmp tree and then diff that.

thockin committed:
Move go-header-file flag to each tool

Also remove the build-tag flag (unused, untested).

This changes some output comments (less redundant).

thockin committed:
Get rid of gengo WithoutDefaultFlagParsing

thockin committed:
Get rid of gengo InputDirs flag

thockin committed:
Get rid of gengo CustomArgs

thockin committed:
The gengo "args" package is gone

thockin committed:
Gengo utils moved

thockin committed:
Renamed gengo Builder -> Parser

thockin committed:
Renamed gengo SourcePath -> Dir

thockin committed:
No need to check nil iterating gengo Inputs

thockin committed:
EOL code-generator/generate-groups scripts

There's no easy way to make it work in the new model.  Callers add flags
to the commandline like:

```
"${CODEGEN_PKG}/generate-internal-groups.sh" "client,conversion,deepcopy,defaulter,informer,lister,openapi" \
  k8s.io/code-generator/examples/apiserver \
  k8s.io/code-generator/examples/apiserver/apis \
  k8s.io/code-generator/examples/apiserver/apis \
  "example:v1 example2:v1 example3.io:v1" \
  --output-base "${SCRIPT_DIR}/../../../.." \
  --go-header-file "${SCRIPT_DIR}/boilerplate.go.txt"
```

To support this we would need to parse these flags in these scripts and
try to adapt them to the new CLI tool interfaces.  Not impossible, but
seeing as we deprecated these a few releases ago AND that we have other
breaking changes to the tools, let's just rip this bandaid off?

thockin committed:
Move import-boss: k/code-generator/cmd -> k/k/cmd

thockin committed:
Get rid of code-generator/cmd/openapi-gen

It's an exact duplicate of k8s.io/kube-openapi/cmd/openapi-gen.

thockin committed:
Fix lint in code-generator/cmd/...

Later commits touch code around here, which triggers verify.  This
commit JUST fixes lint.

thockin committed:
Re-vendor latest kube-openapi and gengo/v2

./hack/pin-dependency.sh k8s.io/kube-openapi latest
./hack/pin-dependency.sh k8s.io/gengo/v2 latest
./hack/update-vendor.sh

thockin committed:
Remove old gengo detritus

thockin committed:
Use go 1.22 for KMS mock

dims commented:
/test 
pull-publishing-bot-validate

dims commented:
/test pull-publishing-bot-validate

thockin committed:
Fix instrumentation tests

1) Fail if we can't read critical env vars
2) Don't rely on KUBE_ROOT env var when `go list` works
3) Don't rely on GOOS env var when `go env` works
4) Don't quietly ignore "can't read" errors

Once we stop ignoring errors, some tests fail for real (and should
always have failed).

The "Imported k8s.io/staging constant" test seems to not be allowed at
all anymore.  Han said to nix it and he'd look async.

Oversall this test is dodgy.  You REALLY can't glue strings together and
expect valid Go module paths.  We should consider a deeper rework.

thockin committed:
Run latest openapi generation

thockin commented:
Pushed with metrics test changes (@logicalhan https://github.com/kubernetes/kubernetes/pull/123529/commits/b881a87245d13ca51cb1f397038237f15139270c) and updated openapi results (@alexzielenski https://github.com/kubernetes/kubernetes/pull/123529/commits/9040c3c0c4f44a228e6aef0b6a98e65527adc5a6)

alexzielenski reviewed (changes requested):
I took a pass over the openapi gen changes. Changes are due to some bug fixes that recently went in:

- https://github.com/kubernetes/kube-openapi/pull/454
- https://github.com/kubernetes/kube-openapi/pull/455

All the enum changes look OK. Mostly lists/maps that were previously ignored by the generator. Some enums also were deduplicated (due to aliasing/re-exporting)

The required changes I reviewed manually to ensure they are all enforced and commented with the validation code. I could not find a validation for one field, so +required should be removed likely from `lastTransitionTime`

https://github.com/kubernetes/kubernetes/blob/1c48b7ec144785b4b2493a01d82e40d6d34ce578/staging/src/k8s.io/api/apiserverinternal/v1alpha1/types.go#L113-L115

thockin commented:
Pushed with update to instrumentation test which passes locally but fails in CI.  Sigh,

thockin commented on pkg/generated/openapi/zz_generated.openapi.go:
```diff
@@ -5238,7 +5243,7 @@ func schema_k8sio_api_apiserverinternal_v1alpha1_StorageVersionCondition(ref com
 					},
 				},
-				Required: []string{"type", "status", "reason"},
+				Required: []string{"type", "status", "lastTransitionTime", "reason", "message"},
```
@jpbetz - how to proceed?  Should we remove `+required`?  This is an internal API and it LOOKS like we always set this field, even if we don't validate it:

```
func setStatusCondition(conditions *[]v1alpha1.StorageVersionCondition, newCondition v1alpha1.StorageVersionCondition,    
    forceTransition bool) {    
    if conditions == nil {    
        return     
    }    
    
    if newCondition.LastTransitionTime.IsZero() {    
        newCondition.LastTransitionTime = metav1.NewTime(time.Now())    
    }    
```

Would it be better to make it actually validated or just document it as not required?

thockin replied:
ping @jpbetz - sorry

liggitt replied:
if we accept and validate requests that don't populate these fields, we should drop the +required annotation and openapi

thockin replied:
We do accept them today.  Given this API is pretty internal I thought we might choose say "we always provide it, so let's start validating".  But dropping it is easy.  Next push.

thockin commented:
Pushed with fixed openapi (thanks @alexzielenski )

thockin commented:
Pushed with down-go-versioned openapi and gengo SHAs and a HEAD rebase.

dims commented:
/approve

dims commented:
/hold for more reviews, please feel free to remove hold when you have enough eyes on it @thockin

thockin commented on hack/lib/golang.sh:
```diff
@@ -344,53 +342,99 @@ readonly KUBE_COVERAGE_INSTRUMENTED_PACKAGES=(
   for target; do
     if [ "${target}" = "ginkgo" ] ||
-       [ "${target}" = "github.com/onsi/ginkgo/ginkgo" ] ||
-       [ "${target}" = "vendor/github.com/onsi/ginkgo/ginkgo" ]; then
```
Are you saying we should keep it?  It's kind of filthy - can we just fix the caller?

liggitt replied:
added in https://github.com/kubernetes/kubernetes/commit/f7427d07e0643f44f85e90645e221e7e117ee36b

we can fix callers ... https://grep.app/search?q=vendor/github.com/onsi/ginkgo/ginkgo&filter[repo.pattern][0]=kubernetes are the ones I see in-project (test-infra, kops, kubeadm, kind are the ones that stand out)

thockin commented:
Rebased

dims commented:
/retest

enj commented on vendor/github.com/ThalesIgnite/crypto11/.gitignore:
We mostly just give up atomic changes to the mock KMS plugin, i.e. any changes made to the plugin would have to be merged untested.  I think this is okay because KMS v2 is GA and we aren't touching most of this code anymore.  Having new in-tree deps would be much worse IMO.

enj commented:
In case GH hides this comment

https://github.com/kubernetes/kubernetes/pull/123529#discussion_r1506148176

thockin commented on hack/verify-openapi-spec.sh:
```diff
@@ -29,7 +29,7 @@ source "${KUBE_ROOT}/hack/lib/init.sh"
 kube::etcd::install
 
-make -C "${KUBE_ROOT}" WHAT=cmd/kube-apiserver
+GOPROXY=off go install ./cmd/kube-apiserver
```
Next push

thockin commented on hack/update-kustomize.sh:
```diff
@@ -62,7 +63,7 @@ git add .
 
 echo -e "\n${color_blue:?}Verifying kubectl kustomize version${color_norm:?}"
-make WHAT=cmd/kubectl
+GOPROXY=off go install ./cmd/kubectl
```
nope - apparently this is not run by the CI verify scripts.  Sigh.

thockin replied:
I will add a commit to make the script run correctly.

liggitt commented on hack/update-vendor.sh:
```diff
@@ -249,9 +250,17 @@ for repo in $(kube::util::list_staging_repos); do
-    kube::util::list_staging_repos \
-        | while read -r X; do echo "-replace k8s.io/${X}=../${X}"; done \
-        | xargs -L 100 go mod edit
+    go mod edit -json \
```
Why was this change required? stomping in replace directives for all the staging repos looks like it should have still worked.

I'd like to keep this as it was if possible, I have a follow-up that cleans up update-vendor to lean more on `go work sync` to push out MVS versions

thockin replied:
I think I can drop this commit - it was really useful when gengo/v2 was a replacement.

It does have the nice effect of removing "this module" from each go.mod.  E.g., reverting this produces changes like:

```
diff --git a/staging/src/k8s.io/api/go.mod b/staging/src/k8s.io/api/go.mod
index 4223dfabea6..75c423e7bda 100644
--- a/staging/src/k8s.io/api/go.mod
+++ b/staging/src/k8s.io/api/go.mod
@@ -36,4 +36,7 @@ require (
        sigs.k8s.io/yaml v1.3.0 // indirect
 )
 
-replace k8s.io/apimachinery => ../apimachinery
+replace (
+       k8s.io/api => ../api
+       k8s.io/apimachinery => ../apimachinery
+)
```

liggitt replied:
yeah, keep status quo for now. I actually have this issue fixed in a staged follow-up commit slashing a lot of obsolete stuff out of update-vendor.sh

thockin replied:
will drop in next push

dims commented:
/triage accepted
/priority important-soon

dims commented on build/build-image/Dockerfile:
```diff
@@ -34,7 +34,7 @@ ENV HOME=/go/src/k8s.io/kubernetes
 
 # Make output from the dockerized build go someplace else
-ENV KUBE_OUTPUT_SUBPATH=_output/dockerized
+ENV KUBE_BUILDNAME=dockerized
```
all github refs are just our copy files. so we should be good ( https://github.com/search?type=code&q=%2FKUBE_OUTPUT_SUBPATH%2F )

thockin replied:
Is this a comment or do I need to do something to resolve this?  Sorry, juggling all the feedback...

dims commented on staging/README.md:
````diff
@@ -57,38 +56,40 @@ import (
   ./hack/update-vendor.sh
 ```
+
+8. Run [`./hack/update-go-workspace.sh`](/hack/update-go-workspace.sh) to add
````
Thanks for the new scripts! will definitely help

dims commented on test/conformance/walk.go:
```diff
@@ -179,12 +179,21 @@ func saveAllTestInfo(dataSet []*ConformanceData) {
-	k8sSRC := *k8sPath + "/_output/local/go/src/k8s.io/kubernetes/"
-	trimmedFile := strings.TrimPrefix(targetFrame.File, k8sSRC)
-	trimmedFile = strings.TrimPrefix(trimmedFile, "/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/")
+	// TODO: This is terribly hacky.  Why can't we rely on the --source flag in
```
+1 for a follow up issue

thockin replied:
https://github.com/kubernetes/kubernetes/issues/123567

dims commented on README.md:
````diff
@@ -43,8 +43,6 @@ If you want to build Kubernetes right away there are two options:
 ```
-mkdir -p $GOPATH/src/k8s.io
-cd $GOPATH/src/k8s.io
 git clone https://github.com/kubernetes/kubernetes
````
Not needing `$GOPATH/src/k8s.io` is AWESOME!

dims commented:
/test pull-publishing-bot-validate

(want to see how the publishing bot reacts ... it's ok if this CI job fails, the periodic job broke somehow and @akhilerm is investigating)

thockin commented on hack/lib/golang.sh:
```diff
@@ -344,53 +342,99 @@ readonly KUBE_COVERAGE_INSTRUMENTED_PACKAGES=(
-      # package prepended.
+      # fully-qualified Go package name.
       echo "${target}"
-    elif [[ "${target}" =~ ^vendor/ ]]; then
-      # Strip vendor/ prefix, since we're building in gomodule mode.
-      echo "${target#"vendor/"}"
```
Good point, will fix

thockin commented on hack/lib/golang.sh:
```diff
@@ -344,53 +342,99 @@ readonly KUBE_COVERAGE_INSTRUMENTED_PACKAGES=(
+    # rude to break things if we don't NEED to.  We can't really test if it
+    # exists or not, because the last element might be an output file (e.g.
+    # bar.test) or even "...".
+    echo "./${target}"
```
if user says "./vendor" we do not strip vendor.

```
$ kube::golang::best_guess_go_targets foo
./foo

$ kube::golang::best_guess_go_targets ./foo
./foo

$ kube::golang::best_guess_go_targets example.com/foo
example.com/foo

$ kube::golang::best_guess_go_targets vendor/foo
foo

$ kube::golang::best_guess_go_targets ginkgo
github.com/onsi/ginkgo/v2/ginkgo

$ kube::golang::best_guess_go_targets github.com/onsi/ginkgo/ginkgo
github.com/onsi/ginkgo/v2/ginkgo

$ kube::golang::best_guess_go_targets vendor/github.com/onsi/ginkgo/ginkgo
github.com/onsi/ginkgo/v2/ginkgo

$ kube::golang::best_guess_go_targets github.com/onsi/ginkgo/v2/ginkgo
github.com/onsi/ginkgo/v2/ginkgo

$ kube::golang::best_guess_go_targets vendor/github.com/onsi/ginkgo/v2/ginkgo
github.com/onsi/ginkgo/v2/ginkgo

$ kube::golang::best_guess_go_targets ./vendor/github.com/onsi/ginkgo/v2/ginkgo
./vendor/github.com/onsi/ginkgo/v2/ginkgo
```

thockin commented on test/integration/auth/rbac_test.go:
```diff
@@ -22,7 +22,7 @@ import (
 	"net/http"
 	"net/http/httputil"
-	gopath "path"
+	"path/filepath"
```
ACK

thockin commented on go.work:
```diff
@@ -19,6 +20,7 @@ use (
 	./staging/src/k8s.io/dynamic-resource-allocation
 	./staging/src/k8s.io/endpointslice
 	./staging/src/k8s.io/kms
+	./staging/src/k8s.io/kms/internal/plugins/_mock
```
I will drop the KMS change and leave that as distinct module and NOT include it in the workspace.  Dropping the code-generator/examples/go.mod means that is the only special-case left.  I think this addresses the comment completely?

liggitt replied:
> I will drop the KMS change and leave that as distinct module and NOT include it in the workspace. Dropping the code-generator/examples/go.mod means that is the only special-case left. I think this addresses the comment completely?

Yes, sgtm

thockin commented on staging/src/k8s.io/code-generator/cmd/import-boss/README.md:
```diff
@@ -94,4 +94,4 @@ inverseRules:
 
 ## Reference
 
-- [import-boss](https://github.com/kubernetes/gengo/tree/master/examples/import-boss)
```
Fixed in a later commit

thockin committed:
Make golang::setup-env turn on workspaces

Both GO111MODULE and GOWORK default to on, so this just unsets them.  We
could set them to explicit values but this seems equivalent and cleaner.

thockin commented on hack/lib/golang.sh:
```diff
@@ -567,6 +563,58 @@ kube::golang::setup_env() {
+  unset GOBIN
+
+  # Explicitly turn on modules.
+  export GO111MODULE=on
```
Will add a commit - PTAL

liggitt replied:
explicitly unsetting sgtm

thockin commented on hack/make-rules/test.sh:
```diff
@@ -257,14 +217,17 @@ runTests() {
   local junit_filename_prefix
   junit_filename_prefix=$(junitFilenamePrefix)
 
-  verifyPathsToPackagesUnderTest "$@"
```
Yes?  Kind of.  A later commit makes sure this runs `go list -find`.  Looking at that, it fails in a weird way if you give it a non-existent input.

```
$ make WHAT=pkg/proxy/IPTABLES
stat /home/thockin/src/kubernetes/pkg/proxy/IPTABLES: directory not found
!!! [0228 12:07:12] Call tree:
!!! [0228 12:07:12]  1: /home/thockin/src/kubernetes/hack/lib/golang.sh:930 kube::golang::normalize_go_targets(...)
!!! [0228 12:07:12]  2: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
+++ [0228 12:07:12] Building go targets for linux/amd64
     (non-static)
no Go files in /home/thockin/src/kubernetes
!!! [0228 12:07:12] Call tree:
!!! [0228 12:07:12]  1: /home/thockin/src/kubernetes/hack/lib/golang.sh:820 kube::golang::build_some_binaries(...)
!!! [0228 12:07:12]  2: /home/thockin/src/kubernetes/hack/lib/golang.sh:974 kube::golang::build_binaries_for_platform(...)
!!! [0228 12:07:12]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0228 12:07:12] Call tree:
!!! [0228 12:07:12]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0228 12:07:12] Call tree:
!!! [0228 12:07:12]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make: *** [Makefile:95: all] Error 1
```

It does say "directory not found" but that's on stderr.  That function ends up emitting nothing on stdout.  Bash, bless its soul, swallows failures and disables errexit in some contexts, which seems to include the `read-array` pattern.  So a bad input ends up with a stderr message and then tries to use "" (aka ".") as an argument to Go commands, resulting in "no Go files in /home/thockin/src/kubernetes".

I make it fail better.

thockin replied:
```
$ make WHAT=nonexist
+++ [0228 12:33:49] Building go targets for linux/amd64
    ./nonexist (non-static)
stat /home/thockin/src/kubernetes/nonexist: directory not found
!!! [0228 12:33:49] Call tree:
!!! [0228 12:33:49]  1: /home/thockin/src/kubernetes/hack/lib/golang.sh:823 kube::golang::build_some_binaries(...)
!!! [0228 12:33:49]  2: /home/thockin/src/kubernetes/hack/lib/golang.sh:976 kube::golang::build_binaries_for_platform(...)
!!! [0228 12:33:49]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0228 12:33:49] Call tree:
!!! [0228 12:33:49]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0228 12:33:49] Call tree:
!!! [0228 12:33:49]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make: *** [Makefile:95: all] Error 1

$ make test WHAT=nonexist
+++ [0228 12:33:53] Set GOMAXPROCS automatically to 6
+++ [0228 12:33:53] Running tests without code coverage and with -race
stat /home/thockin/src/kubernetes/nonexist: directory not found
make: *** [Makefile:190: test] Error 1
```

thockin replied:
new commit in next push

thockin committed:
Make `make` fail better with bad inputs

Ultimately, we're hitting places where bash is just hard to get right.

Instead of:

```
$ make WHAT=nonexist
stat /home/thockin/src/kubernetes/nonexist: directory not found
!!! [0228 12:07:12] Call tree:
!!! [0228 12:07:12]  1: /home/thockin/src/kubernetes/hack/lib/golang.sh:930 kube::golang::normalize_go_targets(...)
!!! [0228 12:07:12]  2: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
+++ [0228 12:07:12] Building go targets for linux/amd64
     (non-static)
no Go files in /home/thockin/src/kubernetes
!!! [0228 12:07:12] Call tree:
!!! [0228 12:07:12]  1: /home/thockin/src/kubernetes/hack/lib/golang.sh:820 kube::golang::build_some_binaries(...)
!!! [0228 12:07:12]  2: /home/thockin/src/kubernetes/hack/lib/golang.sh:974 kube::golang::build_binaries_for_platform(...)
!!! [0228 12:07:12]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0228 12:07:12] Call tree:
!!! [0228 12:07:12]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0228 12:07:12] Call tree:
!!! [0228 12:07:12]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make: *** [Makefile:95: all] Error 1
```

We now get:

```
$ make WHAT=nonexist
+++ [0228 12:33:49] Building go targets for linux/amd64
    ./nonexist (non-static)
stat /home/thockin/src/kubernetes/nonexist: directory not found
!!! [0228 12:33:49] Call tree:
!!! [0228 12:33:49]  1: /home/thockin/src/kubernetes/hack/lib/golang.sh:823 kube::golang::build_some_binaries(...)
!!! [0228 12:33:49]  2: /home/thockin/src/kubernetes/hack/lib/golang.sh:976 kube::golang::build_binaries_for_platform(...)
!!! [0228 12:33:49]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0228 12:33:49] Call tree:
!!! [0228 12:33:49]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0228 12:33:49] Call tree:
!!! [0228 12:33:49]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make: *** [Makefile:95: all] Error 1

$ make test WHAT=nonexist
+++ [0228 12:33:53] Set GOMAXPROCS automatically to 6
+++ [0228 12:33:53] Running tests without code coverage and with -race
stat /home/thockin/src/kubernetes/nonexist: directory not found
make: *** [Makefile:190: test] Error 1
```

thockin commented on hack/lib/init.sh:
```diff
@@ -41,9 +41,33 @@ fi
+# compatible behavior.  Perhaps one day we can remove this.
+if [[ -n "${KUBE_OUTPUT_SUBPATH:-}" ]]; then
+    if [[ -n "${KUBE_BUILDNAME:-}" ]]; then
+        echo "ERROR: \$KUBE_OUTPUT_SUBPATH and \$KUBE_BUILDNAME are mutually exclusive"
```
yes, it's set to "dockerized" in Makefile for `cross-in-a-container`.

I'm not even sure how to run this to fix it - afaict `make cross-in-a-container` is never called?  And it only differs from `make cross` in the definition of KUBE_BUILDNAME=dockerized.

Help me make sense of it?  Should I just turn this into a warning instead?

liggitt replied:
> afaict make cross-in-a-container is never called

ummm.... I'm pretty sure some of the kubernetes release builds use it:

https://grep.app/search?q=cross-in-a-container

welcome to cross-repo implicit make target dependencies

liggitt replied:
> Help me make sense of it? Should I just turn this into a warning instead?

I'm ... not sure? I'm trying to figure out what it meant before if both were set, and what it means now...

liggitt replied:
I'd probably ask sig-release if there's a way to dry-run a release build on this PR the way we do to cut releases

thockin replied:
sorry, you could a bad state at head - kick it off again?

thockin replied:
I pushed a new commit which makes it a warning instead of error.  PTAL

saschagrunert replied:
Yes sure, let's see how that goes: https://console.cloud.google.com/cloud-build/builds/42e8574f-a9c4-467f-9d86-9301d28a96be?project=kubernetes-release-test

dims replied:
Tried locally - `make cross-in-a-container KUBE_DOCKER_IMAGE_TAG=v1.30.0-alpha.4` - in a kubekins docker container.

works fine! https://pastebin.com/zF6b45RG

thockin commented on staging/src/k8s.io/code-generator/cmd/client-gen/generators/fake/generator_fake_for_type.go:
```diff
@@ -153,7 +152,7 @@ func (g *genFakeForType) GenerateType(c *generator.Context, t *types.Type, w io.
 		// Generated apply builder type references required for generated Apply function
 		_, gvString := util.ParsePathGroupVersion(g.inputPackage)
-		m["inputApplyConfig"] = types.Ref(gopath.Join(g.applyConfigurationPackage, gvString), t.Name.Name+"ApplyConfiguration")
+		m["inputApplyConfig"] = types.Ref(filepath.Join(g.applyConfigurationPackage, gvString), t.Name.Name+"ApplyConfiguration")
```
Yes, you are right.  I bothced this part.  The problem is that there are lots of places that do this.  I'll look at it and see if it is less horrible than I imagine.

thockin replied:
Added a commit in next push

thockin committed:
Use path instead of filepath for pkg-paths

thockin commented on pkg/kubelet/util/util_windows_test.go:
apparently, which suggests that some of these tests are NEVER RUN on Windows.

thockin commented:
To not lose it:  https://github.com/kubernetes/kubernetes/pull/123529#discussion_r1504904472 is still unresolved

thockin commented:
Pushed with accumulated fixes.  TTBOMK all unresolved comments are looking for more input before they can be addressed.

thockin commented:
typecheck passes locally and the variable is CLEARLY defined.  Will investigate.

thockin commented:
After a rebase I can repro the typecheck fail but still don't understand it

thockin committed:
Fix a test that doesn't build on Windows

Typecheck should catch this but didn't.  Fortunately I fixed that in an
upcoming commit.

thockin commented:
Cool, cool.  Typecheck at HEAD doesn't catch this but there's a bug in a windows-test that doesn't compile.  Will fix it this PR (first commit)

thockin commented:
Rebased.  Typecheck passes.  Lint should pass.  update-vendor commit is dropped.

4 opens, I think:

https://github.com/kubernetes/kubernetes/pull/123529#discussion_r1504904472

https://github.com/kubernetes/kubernetes/pull/123529#discussion_r1505441549

https://github.com/kubernetes/kubernetes/pull/123529#discussion_r1505505604

https://github.com/kubernetes/kubernetes/pull/123529#discussion_r1506312970

liggitt commented on staging/src/k8s.io/code-generator/examples/go.mod:
```diff
@@ -1,57 +0,0 @@
-go 1.22.0
-
-require (
-	k8s.io/api v0.0.0
-	k8s.io/apimachinery v0.0.0
-	k8s.io/client-go v0.0.0
```
If we have to drop the isolating examples go.mod to make things in this PR work, I can live with doing so.

If we don't have to drop it as part of this PR, I'd rather think about it (and would be interested in @deads2k / @sttts input)

thockin replied:
I will drop 5a88e93c5dc

liggitt commented:
Responded to the open questions I could. Updates for my resolved comments lgtm

thockin committed:
Remove errant +required with no validation code

thockin committed:
Fix up go.mod files after reviews

Because of how the previous 100+ commits were done, so changes snuck
thru that properly belong in earlier commits but it's not really
possible to do that without a lot of effort.

We agreed it was OK to "spackle" these cracks with a final commit.

thockin commented:
Pushed with fixes, more coming

liggitt commented on staging/src/k8s.io/code-generator/cmd/defaulter-gen/output_tests/wholepkg/zz_generated.defaults.go:
```diff
@@ -0,0 +1,85 @@
+package wholepkg
+
+import (
+	runtime "k8s.io/apimachinery/pkg/runtime"
```
pretty sure this is what is picking up the `k8s.io/apimachinery` dependency, not sure if that's intentional

thockin replied:
If this were named testdata, I think it would be OK?  Will try

thockin replied:
Actually no.  Sigh.  These tests used to be triggered by `go generate` in the gengo repo, and are now covered by verify-codegen.  Will restore the publishing bot part.

thockin replied:
next push

liggitt commented on staging/src/k8s.io/code-generator/go.mod:
```diff
@@ -2,45 +2,42 @@
 
-replace k8s.io/code-generator => ../code-generator
+replace (
+	k8s.io/apimachinery => ../apimachinery
```
even leaving examples isolated, some of the output files for the tests reference k8s.io/apimachinery and result in this dependency

if we leave this as-is, we'll need to re-add the publishing bot dependency rule on apimachinery and move code-generator down in the order

thockin committed:
Sort publishing/rules.yaml

Needed because defaulter-gen tests end up depending on apimachinery.

thockin commented:
Rebased, restored publishing-bot sort, fixed codegen (commit is marked WIP until it passes, then I will squash it into the earlier commt).  Still some weird corner case on codegen.

Time to sleep, let's see what happens now.

dims commented:
/retest

thockin commented:
Re-pushed - should pass verify now, but the OUTPUT_PATH topic is still open

dims commented:
/retest

liggitt commented:
a couple questions, but changes lgtm

thockin commented on hack/update-codegen.sh:
```diff
@@ -49,29 +48,33 @@ if [[ "${DBG_CODEGEN}" == 1 ]]; then
     # We need vendor exclusion added at the end since it has to be part of
     # the pathspecs which are specified last.
-    git grep --untracked "$@" ':!:vendor/*'
+    #FIXME: should $@ go at the end?
```
oops.

thockin commented on hack/lib/init.sh:
```diff
@@ -41,9 +35,38 @@ fi
+# Compat: The KUBE_OUTPUT_SUBPATH variable is sometimes passed in by callers.
+# If it is specified, we'll use it as we always have, and fall back on
+# compatible behavior.  Perhaps one day we can remove this.
+_KUBE_OUTPUT_ROOT="${KUBE_ROOT}/_output"
+if [[ -n "${KUBE_OUTPUT_SUBPATH:-}" ]]; then
```
The goal was to make "_output" not be something that people pass in.  KUBE_OUTPUT_SUBPATH is a subpath of the repo, so it's not like you can send output to /tmp or something.  If we thought it was important to send output to ./_otherplace, then my change is wrong.  

I'm just gonna revert it.  It's too loose for my liking, but not worth breaking.  I left the comments but reverted the warning and BUILDNAME stuff.

thockin committed:
Make update-codegen avoid sub-workspaces

thockin commented:
Reverted the BUILDNAME stuff.

Removed FIXME.

I think it's done.  All comments resolved. CI passed last time, so unless I horked something it should pass now.  @liggitt if you are happy me, slap me with some LGTM.  But please actually BE happy.

thockin commented:
/retest

dims commented:
/approve
/lgtm

will wait for @liggitt to remove hold

liggitt commented on hack/lib/init.sh:
```diff
@@ -41,9 +35,21 @@ fi
+#
+# Compat: The KUBE_OUTPUT_SUBPATH variable is sometimes passed in by callers.
+# If it is specified, we'll use it in KUBE_OUTPUT.
+_KUBE_OUTPUT_SUBPATH="${KUBE_OUTPUT_SUBPATH:-_output/local}"
+export KUBE_OUTPUT="${KUBE_ROOT}/${_KUBE_OUTPUT_SUBPATH}"
```
non-blocking nit, could be a follow-up cleanup, but this would get you one less ambient bash var:

```suggestion
export KUBE_OUTPUT="${KUBE_ROOT}/${KUBE_OUTPUT_SUBPATH:-_output/local}"
```

liggitt reviewed (commented):
/lgtm
/approve

One non-blocking nit, unhold at will. May the rebase conflict odds be ever in your favor.

thockin commented:
/unhold

Fire in the hole!

dims commented:
Live picture of @thockin !!

![henry-cavill-as-dc-superhero-superman-flying-6hf1nd1s7h49c0ps](https://github.com/kubernetes/kubernetes/assets/23304/6e2952ec-4005-441d-aa05-9b694f31edbe)

liggitt commented:
![first-try](https://github.com/kubernetes/kubernetes/assets/980082/27b4f2e3-8b50-427f-8f55-628b5372a56b)

dims commented:
@thockin - as soon as `ci-kubernetes-build` and `ci-kubernetes-build-fast` finish running, rest of the periodic jobs pick things up. probably 6-12 hours we'll see some. but more tomorrow. (at least patterns will show up better)

aojea commented:
/honk

inteon commented on staging/src/k8s.io/code-generator/cmd/openapi-gen/main.go:
The commit message linked to this file's removal says "It's an exact duplicate of k8s.io/kube-openapi/cmd/openapi-gen" (https://github.com/kubernetes/kubernetes/pull/123529/commits/1a4450f6ffdad44bfc12fae5911c93e3bd238f0f).
However, there are some differences:
- `--trim-path-prefix` is missing in k8s.io/kube-openapi/cmd/openapi-gen
- `--input-dirs` is missing in k8s.io/kube-openapi/cmd/openapi-gen

@thockin

inteon replied:
Could we revert this commit and un-break projects that rely on these features?

thockin commented:
I cannot figure how to load the github UI for this comment - it just
doesn't work.  So I will reply by email and see if ti works

At the time of removal it was an exact duplicate.  The PR has a lot of
commits :)

`--trim-path-prefix` has no meaning anymore.  The way that packages and
paths are processed in Go workspaces is TOTALLY different than what we were
doing before.  There's no more "read a package name, glue a directory to
the front of it, then strip a prefix" sort of logic.  The bad news is that
it's a break and I don't see how to fix it.  The good news is that the new
approach is better in every way.

`--input-dirs` was "cleaned up" because it makes callers easier and "this
is a breaking change anyway".

I know that my excuses are little comfort - I am sorry that some consumers
are broken, but I spent months trying to find a way to fix this without a
break, and I failed.  If you need to old behavior, you can pin to the older
tags of k8s.io/code-generator until you are ready.

> Message ID: <kubernetes/kubernetes/pull/123529/review/2062602093
> ***@***.***>
>

inteon commented:
> I cannot figure how to load the github UI for this comment - it just
> doesn't work.  So I will reply by email and see if ti works
> 
> At the time of removal it was an exact duplicate.  The PR has a lot of
> commits :)
> 
> `--trim-path-prefix` has no meaning anymore.  The way that packages and
> paths are processed in Go workspaces is TOTALLY different than what we were
> doing before.  There's no more "read a package name, glue a directory to
> the front of it, then strip a prefix" sort of logic.  The bad news is that
> it's a break and I don't see how to fix it.  The good news is that the new
> approach is better in every way.
> 
> `--input-dirs` was "cleaned up" because it makes callers easier and "this
> is a breaking change anyway".
> 
> I know that my excuses are little comfort - I am sorry that some consumers
> are broken, but I spent months trying to find a way to fix this without a
> break, and I failed.  If you need to old behavior, you can pin to the older
> tags of k8s.io/code-generator until you are ready.
> 
> > Message ID: <kubernetes/kubernetes/pull/123529/review/2062602093
> > ***@***.***>
> >
> 

Thanks, I figured out that the flags were changed for all commands (not just openapi-gen) after I posted the question and removed my comment because of that (sorry for the confusion).

For future reference: the flags changed in a breaking way, but all functionality is still there.

pohly commented:
This is not a good place for such questions. Please use Slack the next time.

You are asking for a command in the `k8s.io/kubernetes` package, so of course it needs to pull that package.

Could `import-boss` be moved into a staging directory? Probably, it has no dependencies on k/k. Please create an issue and we can discuss there whether that's worthwhile.
